crates/proof-of-sql/src/sql/postprocessing/evaluate.rs

```diff
@@ -0,0 +1,252 @@
+use super::{PostprocessingError, PostprocessingResult};
+use crate::base::{
+    database::{
+        try_add_subtract_column_types, try_divide_column_types, try_multiply_column_types,
+        ColumnType, OwnedColumn, OwnedTable,
+    },
+    math::decimal::{scale_scalar, try_into_to_scalar, Precision},
+    scalar::Scalar,
+};
+use core::cmp::Ordering;
+use proof_of_sql_parser::{
+    intermediate_ast::{BinaryOperator, Expression, Literal, UnaryOperator},
+    Identifier,
+};
+
+/// Evaluator that evaluates a `proof_of_sql_parser::intermediate_ast::Expression`
+/// on an `OwnedTable` and returns an `OwnedColumn`.
+pub struct PostprocessingEvaluator<S: Scalar> {
+    table: OwnedTable<S>,
+}
+
+impl<S: Scalar> PostprocessingEvaluator<S> {
+    /// Creates a new `PostprocessingEvaluator` with the given owned table.
+    pub fn new(table: &OwnedTable<S>) -> Self {
+        Self {
+            table: table.clone(),
+        }
+    }
+    /// Evaluates the given expression on the owned table and returns the result.
+    pub fn evaluate(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
+        self.visit_expr(expr)
+    }
+}
+
+// Private interface
+impl<S: Scalar> PostprocessingEvaluator<S> {
+    fn visit_expr(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
+        match expr {
+            Expression::Column(identifier) => self.visit_column(*identifier),
+            Expression::Literal(lit) => self.visit_literal(lit),
+            Expression::Binary { op, left, right } => self.visit_binary_expr(*op, left, right),
+            Expression::Unary { op, expr } => self.visit_unary_expr(*op, expr),
+            _ => Err(PostprocessingError::Unsupported(format!(
+                "Expression {:?} is not supported yet",
+                expr
+            ))),
+        }
+    }
+
+    fn visit_column(&self, identifier: Identifier) -> PostprocessingResult<OwnedColumn<S>> {
+        Ok(self
+            .table
+            .inner_table()
+            .get(&identifier)
+            .ok_or(PostprocessingError::ColumnNotFound(identifier.to_string()))?
+            .clone())
```

Comment:
I don't like this `clone`, but I'm not sure if there is an easy way to avoid it.

Reply:
@JayWhite2357 I guess we can leave it as is for now?